### PR TITLE
Only reuse HTTP/2 streams when response writer was completed successfully

### PR DESCRIPTION
**Pull request by JamesNK**
When the HTTP/2 stress suite is run with the latest 5.0 master source code there are some failure results, e.g. 30,000 success, 5 failures. I tracked it down to a race condition between the response being written and the request being aborted.

Currently the stream reuse logic assumes that if the response is successfully written then the stream is safe to reuse. However a race condition occurs if:

1. Server execution ends and HttpProtocol.ProduceEnd is called.
2. Request is aborted after ProduceEnd is called but before the response pipe is completed by ProduceEnd.
3. The abort cancels the response pipe pending read.
4. ReadAsync in `Http2OutputProducer.ProcessDataWrites` gets a canceled result.
5. ProcessDataWrites exits, but because the response is written, the stream is reused.
6. The next request to use the stream will fail

The simple solution is to add an explicit flag at the end of the read loop that checks if the writer completed successfully. That flag is used to decide whether to reuse a stream.

### CODE REVIEW

**Comment by JamesNK** [comment]
I couldn't think of a way to unit test this. There is no hook in ProduceEnd to wait until RST_STREAM from the client is received.

**Comment by jkotalik** [comment]
How does the stress suite look with these changes applied?

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs [hunk 1]
```diff
@@ -156,10 +156,10 @@ public void CompleteStream(bool errored)
...
                 _inputFlowControl.Abort();
 
-                // We only want to reuse a stream that has completely finished writing.
-                // This is to prevent the situation where Http2OutputProducer.ProcessDataWrites
-                // is still running in the background.
-                CanReuse = !_keepAlive && HasResponseCompleted;
+                // We only want to reuse a stream that has completely finished writing
+                // and the request wasn't aborted. This is to ensure
+                // Http2OutputProducer.ProcessDataWrites is in the correct state to be reused.
+                CanReuse = !_keepAlive && HasResponseCompleted && !_connectionAborted;
```
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: To make sure, _connectionAborted, _keepAlive, and HasResponseComplete are not updated in a lock at all or expect to be check while holding a lock?

**Comment by JamesNK** [comment]
> How does the stress suite look with these changes applied?

Before: About 3 failures after 1 minute
After: Zero failures after running for 15 minutes

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs (hunk 1 above)
Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: _connectionAborted might be. I don't know whether it should be inside a lock.

**Comment by BrennanConroy** [comment]
I think your computer is too fast or something, because while this did lower the failure rate I still get failures when running for 5 minutes.

**Comment by BrennanConroy** [comment]
Latest changes ran without error for 15+ minutes in the stress run :+1:

**Comment by JamesNK** [comment]
I ran it for an hour during triage. 0 failures.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs [hunk 2]
```diff
@@ -462,17 +461,9 @@ private async ValueTask ProcessDataWrites()
...
-                    _resetAwaitable.Reset();
-                }
-                else
-                {
-                    // Stream was aborted.
-                    break;
-                }
+                // Wait here for the stream to be reset or disposed.
+                await new ValueTask(_resetAwaitable, _resetAwaitable.Version);
+                _resetAwaitable.Reset();
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs]: Does the fact this is pushed mean that this version of the change also works reliably?

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs]: I ran it for 45 minutes without any failures.

**Review by halter73 — APPROVED** [review]

**Pull request merged**